mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala

```diff
@@ -94,15 +99,19 @@ class HashingTF @Since("1.4.0") (@Since("1.4.0") override val uid: String)
   @Since("2.0.0")
   override def transform(dataset: Dataset[_]): DataFrame = {
     val outputSchema = transformSchema(dataset.schema)
-
-    val hashingTF = new OldHashingTF($(numFeatures)).setBinary($(binary))
-    val func = (terms: Seq[_]) => {
-      val seq = hashingTF.transformImpl(terms)
-      Vectors.sparse(hashingTF.numFeatures, seq)
+    val hashUDF = udf { (terms: Seq[_]) => {
+      val termFrequencies = mutable.HashMap.empty[Int, Double]
+      val setTF =
+        if ($(binary)) (i: Int) => 1.0 else (i: Int) => termFrequencies.getOrElse(i, 0.0) + 1.0
+        terms.foreach { term =>
+          val i = Utils.nonNegativeMod(hashFunc(term), $(numFeatures))
```

Comment:
Can this now call `indexOf`? 
The reason you might not is to avoid accessing `$(numFeatures)` every time, but if so, then that can be a local val.